### PR TITLE
[READY] Add new notebook example to docs about adding Kedro to a notebook project

### PR DESCRIPTION
**Pull request by stichbury**
> **NOTE:** Kedro datasets are moving from `kedro.extras.datasets` to a separate `kedro-datasets` package in
> [`kedro-plugins` repository](https://github.com/kedro-org/kedro-plugins). Any changes to the dataset implementations
> should be done by opening a pull request in that repository.
## Description
Answers #2855 

## Development notes
Introduces a new example with a markdown that mirrors a full runnable notebook in a downloadable directory.
Incrementally adds Kedro features to a notebook project for spaceflights.
[Take a look at the built docs](https://kedro--3128.org.readthedocs.build/en/3128/notebooks_and_ipython/notebook-example/add_kedro_to_a_notebook.html) or review the markdown. PLEASE DO NOT EDIT THE NOTEBOOK!

## Checklist

- [ ] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [ ] Signed off each commit with a [Developer Certificate of Origin (DCO)](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/managing-the-commit-signoff-policy-for-your-repository)
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [ ] Added tests to cover my changes
- [ ] Checked if this change will affect Kedro-Viz, and if so, communicated that with the Viz team

### CODE REVIEW

File: docs/source/notebooks_and_ipython/add_kedro_to_a_notebook.md [hunk 1]
```diff
@@ -0,0 +1,652 @@
...
+random_state = conf_params["model_options"]["random_state"]
+
+```
+
+```python
+def big_function():
+    
+    ####################
+    # Data processing  #
+    ####################
```
Comment (astrojuanlu) [inline docs/source/notebooks_and_ipython/add_kedro_to_a_notebook.md:403]: I like this idea. but I wouldn't even include the full code there, I'd do something like

```python
def big_function():
    ...

big_function()
```

(yes, `...` is valid Python syntax) and skip to the next part. It's mostly for narrative purposes, we don't want users to even write this I think.

**Review by astrojuanlu — COMMENTED** [review]
I really like how this is shaping up, great job @stichbury !

Main comments:

- I really like the flow. One outstanding point is how do we make the user download the code. I propose we tell them to clone a repository, or even open it on Gitpod. Could be `kedro-academy`, could be this one. I'm -0.5 on adding the `docs/examples`, but not completely opposed to it if we think it's the right thing to do.
- I opened a PR with some proposed code changes https://github.com/kedro-org/kedro/pull/3133 including using the config loader with the current directory, so the user doesn't have to create `conf/{base,local}`. We can end there, or we can actually create the dirs, but in a second step IMHO.
- Some parts at the end of the tutorial repeat a lot of code. I'm thinking whether we could assume that the user has the context and we can make the final code blocks more terse (rather than fully self-contained). There's potential of creating some confusion, but at the moment this all feels like too much text.

File: docs/source/notebooks_and_ipython/add_kedro_to_a_notebook.md [hunk 2]
```diff
@@ -0,0 +1,652 @@
...
+r2_score(y_test, y_pred)
+```
+
+## Use Kedro for data processing
+Even if you’re not ready for a full Kedro project, you can still take advantage of its data handling solution in your existing project from within a notebook. This section shows you how.
+
+Kedro’s Data Catalog is a registry of all data sources available for use by the project, and offers a separate place to declare details of the datasets your projects use. Kedro provides [built-in datasets for numerous file types and file systems](https://docs.kedro.org/en/stable/kedro_datasets.html), so you don’t have to write any of the logic for reading or writing data. 
+
+Kedro offers a range of datasets, including CSV, Excel, Parquet, Feather, HDF5, JSON, Pickle, SQL Tables, SQL Queries, Spark DataFrames and more. They are supported with the APIs of pandas, spark, networkx, matplotlib, yaml and more. It relies on `[fsspec](https://filesystem-spec.readthedocs.io/en/latest/)` to read and save data from a variety of data stores including local file systems, network file systems, cloud object stores, and Hadoop. You can pass arguments in to load and save operations, and use versioning and credentials for data access.
```
[the remark this answers is not in the document]
Comment (stichbury) [reply docs/source/notebooks_and_ipython/add_kedro_to_a_notebook.md:105]: It's OK to put the backticks inside the square brackets. I think this was a copy from Notion that obviously does it differently :)

File: docs/source/notebooks_and_ipython/add_kedro_to_a_notebook.md [hunk 3]
```diff
@@ -0,0 +1,652 @@
...
+```python
+X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=3)
+```
+
+[Good software engineering practice](https://towardsdatascience.com/five-software-engineering-principles-for-collaborative-data-science-ab26667a311) suggests that we extract *‘magic numbers’* into named constants, sometimes defined at the top of a file, or outside in a utility file, storing it in a format such as yaml. 
+
+<!--This code needs the user to create a yaml file with this contents. Is there a piece of code we could offer that creates and writes one for them into the appropriate directory to save the reader the manual task? -->
+
+```yml
+# params.yml
```
[the remark this answers is not in the document]
Comment (stichbury) [reply docs/source/notebooks_and_ipython/add_kedro_to_a_notebook.md:165]: Yeah, the thinking was short name -> short file and full name to match our convention -> file we can reuse with Kedro later.

File: docs/source/notebooks_and_ipython/add_kedro_to_a_notebook.md [hunk 4]
```diff
@@ -0,0 +1,652 @@
...
+
+Kedro offers a range of datasets, including CSV, Excel, Parquet, Feather, HDF5, JSON, Pickle, SQL Tables, SQL Queries, Spark DataFrames and more. They are supported with the APIs of pandas, spark, networkx, matplotlib, yaml and more. It relies on `[fsspec](https://filesystem-spec.readthedocs.io/en/latest/)` to read and save data from a variety of data stores including local file systems, network file systems, cloud object stores, and Hadoop. You can pass arguments in to load and save operations, and use versioning and credentials for data access.
+
+To start using the Data Catalog, create a `catalog.yml` in the same folder as your notebook to define datasets that can be used when writing your functions. For example:
+
+
+<!--This code needs the user to create a yaml file with this contents. Is there a piece of code we could offer that creates and writes one for them into the appropriate directory to save the reader the manual task? -->
```
[the remark this answers is not in the document]
Comment (stichbury) [reply docs/source/notebooks_and_ipython/add_kedro_to_a_notebook.md:110]: Yep, agreed. I think we should clone/copy/download the folder as suggested. I want to keep it in the same repo though to ensure that updates are synchronised rather than need two different PRs in different places. Also the `kedro-academy` repo is a bit like the wild west compared to this Kedro repo, and we probably should have an `examples` folder in docs for other example code (should we ever make it).

File: docs/source/notebooks_and_ipython/add_kedro_to_a_notebook.md (hunk 1 above)
Comment (stichbury) [reply docs/source/notebooks_and_ipython/add_kedro_to_a_notebook.md]: This is the only change I've yet to make but I wanted to discuss it because I think it's quite common to do. I've seen a conversation with @noklam somewhere on Slack recently where someone was saying that was the common first step in Kedro-ising for them -- make one big func and then refactor. So I thought it worth including. I don't know though and am happy to be overruled 🤷

File: docs/source/notebooks_and_ipython/notebook-example/add_kedro_to_a_notebook.md [hunk 5]
```diff
@@ -0,0 +1,668 @@
...
+  type: pandas.ExcelDataSet
+  filepath: data/shuttles.xlsx
+```
+
+By using Kedro to load the `catalog.yml` file, you can reference the Data Catalog in your notebook as you load the data for data processing.
+
+```python
+# Using Kedro's DataCatalog
+
+from kedro.io import DataCatalog
```
Comment (merelcht) [inline docs/source/notebooks_and_ipython/notebook-example/add_kedro_to_a_notebook.md:127]: Are we assuming that the user has already installed `kedro` at this point?

File: docs/source/notebooks_and_ipython/notebook-example/add_kedro_to_a_notebook.md [hunk 6]
```diff
@@ -0,0 +1,668 @@
...
+
+## Use Kedro configuration
+Kedro offers a [configuration loader](/kedro.config.ConfigLoader) to abstract loading values from a yaml file. You can use Kedro configuration loading without a full Kedro project and this approach replaces the need to load the configuration file with `yaml.safe_load`.
+
+### Use Kedro's configuration loader to load "magic values"
+To use Kedro's `OmegaConfigLoader` to load `parameters.yml` the code is as follows:
+
+```python
+from kedro.config import OmegaConfigLoader
+from kedro.framework.project import settings
```
Comment (merelcht) [inline docs/source/notebooks_and_ipython/notebook-example/add_kedro_to_a_notebook.md:286]: Why do you need to import `settings` here?

**Review by merelcht — APPROVED** [review]
This is really amazing work @stichbury 🤩 I left two minor comments, but other than that all good from my side.

File: docs/source/notebooks_and_ipython/notebook-example/add_kedro_to_a_notebook.md (hunk 5 above)
Comment (stichbury) [reply docs/source/notebooks_and_ipython/notebook-example/add_kedro_to_a_notebook.md:127]: Good point! Added, but not here, I've put it earlier in the example.

File: docs/source/notebooks_and_ipython/notebook-example/add_kedro_to_a_notebook.md (hunk 6 above)
Comment (stichbury) [reply docs/source/notebooks_and_ipython/notebook-example/add_kedro_to_a_notebook.md]: Good point, that's a legacy line of code, and is now removed.

File: docs/source/notebooks_and_ipython/add_kedro_to_a_notebook.md (hunk 1 above)
Comment (astrojuanlu) [reply docs/source/notebooks_and_ipython/add_kedro_to_a_notebook.md]: Fair enough, not the hill I want to die on 😄

**Review by astrojuanlu — APPROVED** [review]
A bit long, but absolutely wonderful nonetheless 💯 Well done @stichbury !

Two final questions:

1. do we need both the `.ipynb` and the `.md`? I reckon having these makes it easier for distracted users that don't install jupytext to just open the `.ipynb` and go on with their lives. The cost is having the content twice and having to sync it ourselves. Not a big deal, just wanted to make it explicit.
2. do you plan to keep the example under `docs/source/notebook-example`, or is this an intermediate step towards having it on https://github.com/kedro-org/kedro-academy/ ? I think this is excellent tutorial material (welp, I've been delivering similar tutorials all year!) and would be a shame to not have it there. but if there's a rationale to keep it here, then fine.

Apart from these two, LGTM 👍🏽

File: docs/source/notebooks_and_ipython/add_kedro_to_a_notebook.md (hunk 1 above)
Comment (stichbury) [reply docs/source/notebooks_and_ipython/add_kedro_to_a_notebook.md]: Yay, `big_function()` lives to fight another day ⚔️

**Comment by stichbury** [comment]
> A bit long, but absolutely wonderful nonetheless 💯 Well done @stichbury !
Thanks!

> Two final questions:
> 
> 1. do we need both the `.ipynb` and the `.md`? I reckon having these makes it easier for distracted users that don't install jupytext to just open the `.ipynb` and go on with their lives. The cost is having the content twice and having to sync it ourselves. Not a big deal, just wanted to make it explicit.

Yes, I wrestled over this but I don't want to force people to use jupytext since this is onboarding type content. The goal is to have a notebook ready to go, but it does come at the expense of us having to understand how to work with the two files. I did write a README for this but removed it. Maybe I'll write a comment in the markdown file to explain how to make updates before I merge this.

> 2. do you plan to keep the example under `docs/source/notebook-example`, or is this an intermediate step towards having it on https://github.com/kedro-org/kedro-academy/ ? I think this is excellent tutorial material (welp, I've been delivering similar tutorials all year!) and would be a shame to not have it there. but if there's a rationale to keep it here, then fine.
> 
I want to keep this in the docs repo. Reason being that otherwise part 1. becomes difficult. We have to copy the markdown across repos, or the notebook, and that just gets annoying. At least with both in the same directory the code and example are side by side, and to make that happen and be able to build the docs, it has to be here. Also, I looked at the `academy` repo and it's not really ideal, it's a bit of a wilderness for a newbie Kedro user. I think we should probably have an examples folder in the docs and I tried it for this, but as per my comment above about putting both files together, it wasn't ideal in this instance, so I moved them to sit under the notebooks section of the docs. 

Thanks for asking: making this explicit is a good decision record.
